src/trace/implementations/mod.rs

```diff
@@ -167,3 +168,251 @@ impl<T: Columnation> RetainFrom<T> for TimelyStack<T> {
+        /// count the number of elements in time logarithmic in the result.
+        fn advance<F: Fn(&Self::Item)->bool>(&self, start: usize, end: usize, function: F) -> usize {
+
+            let small_limit = 8;
+
+            // Exponential seach if the answer isn't within `small_limit`.
+            if end > start + small_limit && function(self.index(start + small_limit)) {
+
+                // start with no advance
+                let mut index = small_limit + 1;
+                if start + index < end && function(self.index(start + index)) {
+
+                    // advance in exponentially growing steps.
+                    let mut step = 1;
+                    while start + index + step < end && function(self.index(start + index + step)) {
+                        index += step;
+                        step = step << 1;
+                    }
+
+                    // advance in exponentially shrinking steps.
+                    step = step >> 1;
+                    while step > 0 {
+                        if start + index + step < end && function(self.index(start + index + step)) {
+                            index += step;
+                        }
+                        step = step >> 1;
+                    }
+
+                    index += 1;
+                }
+
+                index
+            }
+            else {
+                let limit = std::cmp::min(end, start + small_limit);
+                (start .. limit).filter(|x| function(self.index(*x))).count()
+            }
+        }
+    }
+
+    // All `T: Clone` also implement `ToOwned<Owned = T>`, but without the constraint Rust
+    // struggles to understand why the owned type must be `T` (i.e. the one blanket impl).
+    impl<T: Clone + ToOwned<Owned = T>> BatchContainer for Vec<T> {
+        type Item = T;
+        fn push(&mut self, item: T) {
+            self.push(item);
+        }
+        fn copy(&mut self, item: &T) {
+            self.push(item.clone());
+        }
+        fn copy_slice(&mut self, slice: &[T]) where T: Sized {
+            self.extend_from_slice(slice);
+        }
+        fn copy_range(&mut self, other: &Self, start: usize, end: usize) {
+            self.extend_from_slice(&other[start .. end]);
+        }
+        fn with_capacity(size: usize) -> Self {
+            Vec::with_capacity(size)
+        }
+        fn reserve(&mut self, additional: usize) {
+            self.reserve(additional);
+        }
+        fn merge_capacity(cont1: &Self, cont2: &Self) -> Self {
+            Vec::with_capacity(cont1.len() + cont2.len())
+        }
+        fn index(&self, index: usize) -> &Self::Item {
+            &self[index]
+        }
+        fn len(&self) -> usize {
+            self[..].len()
+        }
+    }
+
+    // The `ToOwned` requirement exists to satisfy `self.reserve_items`, who must for now
+    // be presented with the actual contained type, rather than a type that borrows into it.
```

Comment:
We should revisit the type requirements on region and timely stack!

Reply:
Right! I'm not sure if it really helps here, but e.g. if I have a bunch of `Vec<u8>` and you are able to only maintain `[u8]` shaped data, that could help. Or we could conclude that this type does not do that, and not be too stressed! :D